hera_pspec/grouping.py

```diff
@@ -288,74 +307,135 @@ def average_spectra(uvp_in, blpair_groups=None, time_avg=False,
+                            bpg_data.append(data * error_weights)
+                            bpg_wgts.append(wgts * error_weights[:, :, None])
+                            bpg_ints.append(ints * error_weights)
+                            bpg_nsmp.append(nsmp)
+                            for stat in stat_l:
+                                bpg_stats[stat].append(errws[stat]*error_weights**2)
+                            w_list.append(error_weights)
+                            if store_cov:
+                                bpg_cov.append(cov * error_weights[:, :, None]**2)
+
+                    else:
+                    # Otherwise all arrays are averaged in a way weighted by the integration time
+                        w = (ints * np.sqrt(nsmp))
+                        # shaoe of w: (Ntimes, 1)
+                        # Get error bar weights and set invalid ones to zero.
+                        errws = {}
+                        for stat in stat_l:
+                            errs = uvp.get_stats(stat, (spw, blp, p)).copy()
+                            no_data = np.isnan(errs)
+                            errs[~no_data] = 1. / errs[~no_data] ** 2.
+                            errs[no_data] = 0.
+                            errws[stat] = errs
+                    
+                        # Take time average if desired
+                        if time_avg:
+                          data = (np.sum(data * w, axis=0) \
+                               / np.sum(w, axis=0).clip(1e-10, np.inf))[None]
+                          wgts = (np.sum(wgts * w[:, None], axis=0) \
+                               / np.sum(w, axis=0).clip(1e-10, np.inf)[:, None])[None]
+                          ints = (np.sum(ints * w, axis=0) \
+                               / np.sum(w, axis=0).clip(1e-10, np.inf))[None]
+                          nsmp = np.sum(nsmp, axis=0)[None]
+                          if store_cov:
+                              cov = (np.sum(cov * w[:, None]**2, axis=0) \
+                              / (np.sum(w,axis=0)**2).clip(1e-10, np.inf))[None]
+
+                          for stat in stat_l:
+                            errws[stat] = np.sum(errws[stat], axis=0)[None]
+                          w = np.sum(w, axis=0)[None]
+                        # Add multiple copies of data for each baseline according
+                        # to the weighting/multiplicity
+                        for m in range(int(blpg_wgts[k])):
+                            bpg_data.append(data * w)
+                            bpg_wgts.append(wgts * w[:, None])
+                            bpg_ints.append(ints * w)
+                            bpg_nsmp.append(nsmp)
+                            [bpg_stats[stat].append(errws[stat]) for stat in stat_l]
+                            w_list.append(w)
+                            if store_cov:
+                                bpg_cov.append(cov * w[:, None]**2)
```

Comment:
Are these two blocks of code in the `if` statement almost the same, except that the first block uses the specified `error_weights` and the second uses the integration time as the weights? 

If (and only if) possible, it would be good to just have one block of code that does the averaging/time averaging/multiplicity (to avoid having two copies of very similar code in both parts of the `if` statement). The weights could be calculated in a shorter `if` statement above this block.

Reply:
They are not the same on how to average the error fields. In the original code, each error field is weighted by itself (1/sigma_i^2), while other arrays are weighted by the integration time; In the new code, every array is weighted by the specific error. I hope we could have more clarification here.

Reply:
OK, if it can't be simplified, then that's fine. (May be worth mentioning in an additional in-line comment though.)

Reply:
From the discussion on today's pspec telcon, people agree us to do the average in a consistent way in which every array is either weighted by the integration time or the error bars.